File: src/main/java/com/busted_moments/client/screen/territories/search/criteria/IsCriteria.java
package com.busted_moments.client.screen.territories.search.criteria;

import com.busted_moments.client.models.territory.eco.TerritoryEco;
import com.busted_moments.client.screen.territories.search.Criteria;
import com.busted_moments.client.screen.territories.search.Operator;
import com.busted_moments.client.screen.territories.search.Operators;
import com.busted_moments.core.util.EnumUtil;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Stream;

@Criteria.With("is")
@Operators(Operator.IS)
public class IsCriteria extends Criteria {
   private static final List<String> SUGGESTIONS = Stream.of(Type.values())
           .map(Enum::name)
           .map(String::toLowerCase)
           .toList();

   public IsCriteria(Operator operator) throws UnsupportedOperationException {
      super(operator);
   }

   @Override
   public Compiled compile(String value) {
      return new Compiled(this, value, EnumUtil.valueOf(value, Type.class).orElseThrow());
   }

   @Override
   public List<String> suggestions() {
      return SUGGESTIONS;
   }

   private enum Type implements Predicate<TerritoryEco> {
      HQ(TerritoryEco::isHQ),
      CONNECTION(TerritoryEco::isConnection),
      EXTERNAL(TerritoryEco::isExternal);

      private final Predicate<TerritoryEco> predicate;

      Type(Predicate<TerritoryEco> predicate) {
         this.predicate = predicate;
      }

      @Override
      public boolean test(TerritoryEco territoryEco) {
         return predicate.test(territoryEco);
      }
   }
}

File: src/main/kotlin/com/busted_moments/client/screens/territories/search/criteria/IsCriteria.kt
package com.busted_moments.client.screens.territories.search.criteria

import com.busted_moments.client.commands.args.war.WarFilter
import com.busted_moments.client.models.territories.eco.TerritoryData
import com.busted_moments.client.screens.territories.search.Criteria
import com.mojang.brigadier.StringReader
import java.util.function.Predicate

class IsCriteria(
    val type: Type
) : Criteria {
    override fun test(t: TerritoryData): Boolean =
        type.test(t)

    companion object : Criteria.Builder {
        override val names: Array<String>
            get() = arrayOf("is")
        override val operators: Array<WarFilter.Operator>
            get() = arrayOf(WarFilter.Operator.IS)

        override val suggestions: Array<String>
            get() = arrayOf("hq", "connections", "conn", "external", "ext")

        override fun parse(reader: StringReader, operator: WarFilter.Operator): Criteria =
            IsCriteria(enumValueOf(reader.readString().uppercase()))
    }

    enum class Type : Predicate<TerritoryData> {
        HQ {
            override fun test(t: TerritoryData): Boolean =
                t.hq
        },
        CONNECTION {
            override fun test(t: TerritoryData): Boolean =
                t.absoluteRoute == null || t.absoluteRoute!!.size <= 1

        },
        CONN {
            override fun test(t: TerritoryData): Boolean =
                CONNECTION.test(t)
        },
        EXTERNAL {
            override fun test(t: TerritoryData): Boolean =
               t.absoluteRoute == null || t.absoluteRoute!!.size <= 3
        },
        EXT {
            override fun test(t: TerritoryData): Boolean =
                EXTERNAL.test(t)
        }
    }
}